README.rst
===============================
phonevalidator
===============================

.. image:: https://img.shields.io/travis/m-housh/phonevalidator.svg
        :target: https://travis-ci.org/m-housh/phonevalidator

.. image:: https://coveralls.io/repos/github/m-housh/phonevalidator/badge.svg?branch=master
        :target: https://coveralls.io/github/m-housh/phonevalidator?branch=master


Custom `cerberus.Validator` for phone numbers


* Free software: MIT license


Features
--------

* Custom validation mixin for `Cerberus`_ or an `Eve`_ API.  
  Uses the `phonenumbers`_ package for validation's and formatting.
  Tested under Python 2.7, Python 3.3, Python 3.4, and Python 3.5.

Basic Usage
-----------
-------

Cerberus Usage:
===============
.. code-block:: python

    from phonevalidator import Validator

    schema = {
        'phone': {
            'type': 'phonenumber',
            'formatPhoneNumber': True,
            'phoneNumberFormat': 'NATIONAL',
            'region': 'US',
        },
    }

    doc = {'phone': '5135555555'}

    v = Validator(schema)
    v.validate(doc)
    # True
    v.document
    # {'phone': '(513) 555-5555'}

    doc = {'phone': 'gibberish'}
    v.validate(doc)
    # False

Eve Usage:
==========
.. code-block:: python

    from phonevalidator import ValidatorMixin
    from eve.io.mongo import Validator
    from eve import Eve


    class MyValidator(Validator, ValidatorMixin):
        """ My custom validator that includes phone number 
        validations.
        """
        pass


    app = Eve(validator=MyValidtor)
    ...


Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

This package also use the `phonenumbers`_ package which is licensed under
the `Apache2`_ license.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Cerberus`:  http://docs.python-cerberus.org
.. _`Eve`: http://python-eve.org
.. _`phonenumbers`:  https://github.com/daviddrysdale/python-phonenumbers
.. _`Apache2`: http://www.apache.org/licenses/LICENSE-2.0


HISTORY.rst
=======
History
=======

0.0.1 (2016-07-22)
------------------

* First release on PyPI.


tests/test_phonevalidator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_phonevalidator
----------------------------------

Tests for `phonevalidator` module.
"""

import pytest
import os
from phonevalidator import Validator
from phonenumbers import PhoneNumberFormat


@pytest.fixture
def base_schema():
    return {'phone': {'type': 'phonenumber', 'region': 'US'}}


def test_Validator_init():
    v = Validator(formatter='INTERNATIONAL', region='AS')
    assert v.formatter == PhoneNumberFormat.INTERNATIONAL
    assert v.region == 'AS'

    v = Validator()
    assert v.formatter is None
    assert v.region is None


def test_Validator_default_region():
    v = Validator()
    assert v._default_region() == 'US'

    os.environ['DEFAULT_PHONE_REGION'] = 'AS'
    assert v._default_region() == 'AS'


def test_Validator_set_formatter():
    v = Validator()
    assert v.formatter is None
    v._set_formatter('NOT_VALID')
    assert v.formatter == PhoneNumberFormat.NATIONAL
    v.formatter = None
    v._set_formatter()
    assert v.formatter == PhoneNumberFormat.NATIONAL


def test_Validator_set_default_formatter():
    v = Validator()
    v._set_default_formatter()
    assert v.formatter == PhoneNumberFormat.NATIONAL
    os.environ['DEFAULT_PHONE_FORMAT'] = 'INTERNATIONAL'
    v._set_default_formatter()
    assert v.formatter == PhoneNumberFormat.INTERNATIONAL


def test_Validator_is_valid_region():
    v = Validator()
    assert v._is_valid_region(None) is False
    assert v._is_valid_region('us') is True


def test_phonenumber_fail(base_schema):
    doc = {'phone': 'gibberish'}
    v = Validator(base_schema)
    assert v.validate(doc) is False

    doc['phone'] = '+41513555'
    assert v.validate(doc) is False


def test_phone_number_valid(base_schema):
    base_schema['phone']['region'] = 'US'
    doc = {'phone': '5135555555'}
    v = Validator(base_schema)
    assert v.validate(doc)
    assert v.document['phone'] == '5135555555'


def test_phoneNumberFormat_fail(base_schema):
    base_schema['phone'].update({
        'phoneNumberFormat': 'InvalidFormat',
        'formatPhoneNumber': True,
        'region': 'US',
    })
    doc = {'phone': '5135555555'}
    v = Validator(base_schema)
    assert v.validate(doc) is False


def test_region_fail(base_schema):
    # non valid region
    schema = {'phone': {'type': 'phonenumber', 'region': 'XYZ'}}
    doc = {'phone': '5135555555'}
    v = Validator(schema)
    assert v.validate(doc) is False


def test_phone_gets_formatted(base_schema):
    base_schema['phone'].update({
        'formatPhoneNumber': True,
        'phoneNumberFormat': 'NATIONAL',
        'region': 'US'})
    doc = {'phone': '5135555555'}
    v = Validator(base_schema)
    assert v.validate(doc)
    assert v.document['phone'] == '(513) 555-5555'


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'm-housh/python_boilerplate'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


phonevalidator/phonevalidator.py
# -*- coding: utf-8 -*-

from os import environ
from phonenumbers import (
    format_number,
    is_valid_number,
    parse,
    NumberParseException,
    PhoneNumberFormat,
    SUPPORTED_REGIONS,
)


class ValidatorMixin(object):
    """ A custom cerberus.Validator subclass adding the `phonenumber` constraint
    to Cerberus validation's.
    """
    '''
    def __init__(self, *args, **kwargs):
        super(ValidatorMixin, self).__init__()
        self.formatter = None
        self.region = None

        if formatter is not None:
            # set formatter for this instance to the formatter
            # passed in if valid, if not self.formatter will
            # get set to the default (`PhoneNumberFormat.NATIONAL`)
            self._set_formatter(
                formatter=formatter
            )

        if region is not None:
            # set the region for this instance to validate
            # phone numbers against if valid, if not valid
            # the self.region will get set to the default
            # ('US')
            self._set_region(
                region=region
            )
        '''

    def _default_region(self):
        """ Return's a default region string.

            We will try to get a region from the environment variable
            'DEFAULT_PHONE_REGION' if not found or not valid, then
            we default to 'US'
        """
        region = environ.get('DEFAULT_PHONE_REGION')
        if region and region.upper() in SUPPORTED_REGIONS:
            return region.upper()
        return 'US'

    def _is_valid_region(self, region):
        """ Return's True if region is valid, False otherwise
        """
        if isinstance(region, str) and region.upper() in SUPPORTED_REGIONS:
            return True
        return False

    def _set_region(self, region=None):
        """ Set's a region to use to validate phone numbers for this
            instance.
        """
        if self._is_valid_region(region):
            self.region = region.upper()
        else:
            self.region = self._default_region()

    def _set_default_formatter(self):
        """ Try's to get a formatter string from the environment
            or defaults to `phonenumbers.PhoneNumberFormat.NATIONAL`
        """
        env_format = environ.get('DEFAULT_PHONE_FORMAT')
        if env_format is not None:
            self._set_formatter(
                formatter=env_format
            )
        else:
            self.formatter = PhoneNumberFormat.NATIONAL

    def _set_formatter(self, formatter=None):
        """ Set's formatter for this instance, or defaults to
            `phonenumbers.PhoneNumberFormat.NATIONAL`
        """
        if formatter is None or not isinstance(formatter, str):
            self._set_default_formatter()
        else:
            self.formatter = getattr(
                PhoneNumberFormat,
                formatter.upper(),
                PhoneNumberFormat.NATIONAL
            )

    def _validate_formatPhoneNumber(self, formatPhoneNumber, field, value):
        """ Fake validate function to let cerberus accept "formatPhoneNumber"
            as a keyword in the schema.
        """
        pass

    def _validate_phoneNumberFormat(self, phoneNumberFormat, field, value):
        """ Validates a phoneNumberFormat for a phone number.

            :param phoneNumberFormat:  a string for accepted format.
            :accepted formats: ['NATIONAL',
                                'INTERNATIONAL',
                                'RFC3966',
                                'E164'
                                ]

        """
        keys = PhoneNumberFormat.__dict__.keys()
        valids = [key for key in keys if not key.startswith('_')]

        if phoneNumberFormat.upper() not in valids:
            self._error(field,
                        'Not a valid phone number format: {}'.format(value))

    def _validate_region(self, region, field, value):
        """ Validates a region for a phone number.

            :param region:  a region.
        """
        if self._is_valid_region(region) is False:
            self._error(field,
                        'Region not valid: {}'.format(region))

    def _validate_type_phonenumber(self, field, value):
        """ Validates a phone number is valid. Optionally formatting the number.

            :param field:  field name.
            :param value:  field value.
        """
        # get the region from schema for this field or use default
        self._set_region(self.schema[field].get('region'))
        try:
            phone_number = parse(value, self.region)

            # check that it's valid number
            if not is_valid_number(phone_number):
                self._error(field, 'Phone Number not valid: {}'.format(value))
            elif self.schema[field].get('formatPhoneNumber'):
                # if the schema's 'formatPhoneNumber' is set to True,
                # format the phone number using a formatter derived from
                # the schema's 'phoneNumberFormat' value, next checks the
                # environmen variable 'PHONE_NUMBER_FORMAT',
                # or defaults to 'NATIONAL'.
                formatter = self.schema[field].get('phoneNumberFormat')
                self._set_formatter(
                    formatter=formatter
                )
                self.document[field] = format_number(phone_number,
                                                     self.formatter)
        except NumberParseException:
            self._error(field, 'Phone Number not valid: {}'.format(value))


phonevalidator/__init__.py
# -*- coding: utf-8 -*-

from .phonevalidator import ValidatorMixin
from cerberus import Validator as SuperValidator


class Validator(SuperValidator, ValidatorMixin):
    """ Extends `cerberus.Validator` and adds the `phonenumber` constraint
    to Cerburus validation's.

    :Example:

    .. code-block:: python

        >>> from phonevalidator import Validator
        >>> schema = {
        ...     'phone': {
        ...         'type': 'phonenumber',
        ...         'formatPhoneNumber': True,
        ...         'phoneNumberFormat': 'NATIONAL',
        ...         'region': 'US'
        ...     }
        ... }
        >>> doc = {'phone': '5135555555'}
        >>> v = Validator(schema)
        >>> v.validate(doc)
        True
        >>> v.document
        {'phone': '(513) 555-5555'}
        >>> doc = {'phone': 'gibberish'}
        >>> v.validate(doc)
        False


    """
    pass

__author__ = 'Michael Housh'
__email__ = 'mhoush@houshhomeenergy.com'
__version__ = '1.1.0'
__all__ = ['Validator', 'ValidatorMixin']
